internal/compiler/lookup.rs

```diff
@@ -219,90 +225,134 @@ impl LookupObject for IdLookup {
+                    if let Some(r) = visit_entry(
                         &repeated.model_data_id,
-                        Expression::RepeaterIndexReference { element: Rc::downgrade(elem) }.into(),
+                        Expression::RepeaterModelReference { element: Rc::downgrade(elem) }.into(),
                     ) {
                         return Some(r);
                     }
                 }
             }
 
-            if let Some(r) = elem.for_each_entry(ctx, f) {
-                return Some(r);
+            if is_legacy {
+                if elem.borrow().repeated.is_some()
+                    || idx == 0
+                    || idx == ctx.component_scope.len() - 1
+                {
+                    if let Some(r) = visit_legacy_scope(elem) {
+                        return Some(r);
+                    }
+                }
+            } else {
+                if let Some(r) = visit_scope(elem) {
+                    return Some(r);
+                }
             }
         }
         None
     }
+}
+impl LookupObject for InScopeLookup {
+    fn for_each_entry<R>(
+        &self,
+        ctx: &LookupCtx,
+        f: &mut impl FnMut(&str, LookupResult) -> Option<R>,
+    ) -> Option<R> {
+        let f = RefCell::new(f);
+        Self::visit_scope(
+            ctx,
+            |str, r| f.borrow_mut()(str, r),
+            |elem| elem.for_each_entry(ctx, *f.borrow_mut()),
+            |scope| {
+                for (name, prop) in &scope.borrow().property_declarations {
+                    let e = expression_from_reference(
+                        NamedReference::new(scope, name),
+                        &prop.property_type,
+                    );
+                    if let Some(r) = f.borrow_mut()(name, e.into()) {
+                        return Some(r);
+                    }
+                }
+                None
+            },
+        )
+    }
 
     fn lookup(&self, ctx: &LookupCtx, name: &str) -> Option<LookupResult> {
         if name.is_empty() {
             return None;
         }
-        for elem in ctx.component_scope.iter().rev() {
-            if let Some(repeated) = &elem.borrow().repeated {
-                if repeated.index_id == name {
-                    return Some(LookupResult::from(Expression::RepeaterIndexReference {
-                        element: Rc::downgrade(elem),
-                    }));
-                }
-                if repeated.model_data_id == name {
-                    return Some(LookupResult::from(Expression::RepeaterModelReference {
-                        element: Rc::downgrade(elem),
-                    }));
-                }
-            }
-
-            if let Some(r) = elem.lookup(ctx, name) {
-                return Some(r);
-            }
-        }
-        None
+        Self::visit_scope(
+            ctx,
+            |str, r| (str == name).then(|| r),
+            |elem| elem.lookup(ctx, name),
+            |scope| {
```

Comment:
Same here maybe, since there is a `ComponentScope` type as well that this could be mistaken as (although the use of `property_declarations` is typically associative with `ElementRc`.